File: src/_03_BehavioralPattern/_03_08_State/java/after/DraftState.java
package _03_BehavioralPattern._03_08_State.java.after;

public class DraftState implements State {

  private OnlineCourse onlineCourse;

  public DraftState(OnlineCourse onlineCourse) {
    this.onlineCourse = onlineCourse;
  }

  @Override
  public void addStudent(Student student) {
    this.onlineCourse.getStudents().add(student);

    if (this.onlineCourse.getStudents().size() > 1) {
      this.onlineCourse.changeState(new PrivateState(this.onlineCourse));
    }
  }

  @Override
  public void addReview(Student student, String review) {
    throw new UnsupportedOperationException("드래프트 상태에서는 리뷰를 남길 수 없습니다.");
  }
  
  @Override
  public String toString() {
    return "<Draft 상태>";
  }
  
}

File: src/_03_BehavioralPattern/_03_08_State/DraftState.js
import State from "./State.js";
import PrivateState from "./PrivateState.js";

/**
 * @typedef { import("./OnlineCourse").default } OnlineCourse
 * @typedef { import("./Student").default } Student
 */

export default class DraftState extends State {
  /** @type { OnlineCourse } */
  _onlineCourse;

  /** @param { OnlineCourse } onlineCourse */
  constructor(onlineCourse) {
    super();
    this._onlineCourse = onlineCourse;
  }

  /**
   * @override
   * @param { Student } student 
   */
  addStudent(student) {
    this._onlineCourse.getStudents().push(student);

    if (this._onlineCourse.getStudents().length > 1) {
      this._onlineCourse.changeState(new PrivateState(this._onlineCourse));
    }
  }

  /**
   * @override
   * @param { Student } _student 
   * @param { string } _review 
   */
  addReview(_student, _review) {
    throw new Error("드래프트 상태에서는 리뷰를 남길 수 없습니다.");
  }
}